File: util-c/src/util/automaton.c
#include <util/automaton.h>
#include <stdio.h>

static util_error fire_action( util_automaton * This, const util_automaton_state_action * actions, size_t count ) {
   for( unsigned i = 0; i < count; ++i ) {
      const util_automaton_state_action * action = actions + i;
      if( action->state == This->current ) {
         (action->action)( This->actor );
         return UTIL_NO_ERROR;
      }
   }
   return UTIL_NOT_APPLICABLE;
}

util_error util_automaton_process( util_automaton * This, int event ) {
   if( ! This ) {
      return UTIL_NULL_ARG;
   }
   for( unsigned i = 0; i < This->shortcuts_count; ++i ) {
      const util_shortcut * shortcut = This->shortcuts + i;
      if( shortcut->event == event ) {
         fire_action( This, This->on_exit, This->on_exit_count );
         This->current = shortcut->next;
         fire_action( This, This->on_entry, This->on_entry_count );
         return UTIL_NO_ERROR;
      }
   }
   for( unsigned i = 0; i < This->arcs_count; ++i ) {
      const util_transition * arc = This->arcs + i;
      if( arc->current == This->current && arc->event == event ) {
         fire_action( This, This->on_exit, This->on_exit_count );
         This->current = arc->next;
         fire_action( This, This->on_entry, This->on_entry_count );
         return UTIL_NO_ERROR;
      }
   }
   fprintf( stderr, "util_automaton_process|ignored event %d\n", event );
   return UTIL_NOT_FOUND;
}

File: udt-java/src/dab/udt/Automaton.java
package dab.udt;

import java.util.HashMap;
import java.util.Map;

public class Automaton<S, E> {

   private final Map<S, Map< E, S>> _arcs      = new HashMap<>();
   private final Map< E, S>         _shortcuts = new HashMap<>();
   private /* */ S                  _current;

   public Automaton( S initial, Arc<S, E>[] arcs, Shortcut<S, E>[] shortcuts ) {
      _current = initial;
      for( final Arc<S, E> arc : arcs ) {
         Map< E, S> transition = _arcs.get( arc.current );
         if( transition == null ) {
            _arcs.put( arc.current, transition = new HashMap<>());
         }
         transition.put( arc.event, arc.futur );
      }
      for( final Shortcut<S, E> s : shortcuts ) {
         _shortcuts.put( s.event, s.futur );
      }
   }

   static final class Arc<S, E> {

      Arc( S c, E e, S f ) {
         current = c;
         event   = e;
         futur   = f;
      }

      private final S current;
      private final E event;
      private final S futur;
   }

   static final class Shortcut<S, E> {

      Shortcut( E e, S f ) {
         event = e;
         futur = f;
      }

      private final E event;
      private final S futur;
   }

   public boolean process( Event event ) {
      S futur = _shortcuts.get( event );
      if( futur != null ) {
         _current = futur;
         return true;
      }
      final Map<E, S> transition = _arcs.get( _current );
      if( transition != null ) {
         futur = transition.get( event );
         if( futur != null ) {
            _current = futur;
            return true;
         }
      }
      return false;
   }

   public S getCurrentState() {
      return _current;
   }
}
